// @wdio/devtools-service/build/auditor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _weightedMean = _interopRequireDefault(require("weighted-mean"));

var _diagnostics = _interopRequireDefault(require("lighthouse/lighthouse-core/audits/diagnostics"));

var _mainthreadWorkBreakdown = _interopRequireDefault(require("lighthouse/lighthouse-core/audits/mainthread-work-breakdown"));

var _metrics = _interopRequireDefault(require("lighthouse/lighthouse-core/audits/metrics"));

var _timeToFirstByte = _interopRequireDefault(require("lighthouse/lighthouse-core/audits/time-to-first-byte"));

var _utils = require("./utils");

var _logger = _interopRequireDefault(require("@wdio/logger"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const log = (0, _logger.default)('@wdio/devtools-service:Auditor');
const METRIC_SCORING = {
  FMP: [4000, 1600],
  FI: [10000, 1700],
  CI: [10000, 1700],
  SI: [5500, 1250],
  EIL: [100, 50]
};
const WEIGHT_WITHIN_CATEGORY = {
  FMP: 5,
  FI: 5,
  CI: 5,
  SI: 1,
  EIL: 1
};
const SHARED_AUDIT_CONTEXT = {
  settings: {
    throttlingMethod: 'devtools'
  },
  LighthouseRunWarnings: false,
  computedCache: new Map()
};

class Auditor {
  constructor(traceLogs, devtoolsLogs) {
    this.devtoolsLogs = devtoolsLogs;
    this.traceLogs = traceLogs;
    this.url = traceLogs.pageUrl;
    this.loaderId = traceLogs.loaderId;
  }

  _audit(AUDIT, params = {}) {
    const auditContext = _objectSpread({
      options: _objectSpread({}, AUDIT.defaultOptions)
    }, SHARED_AUDIT_CONTEXT);

    try {
      return AUDIT.audit(_objectSpread({
        traces: {
          defaultPass: this.traceLogs
        },
        devtoolsLogs: {
          defaultPass: this.devtoolsLogs
        }
      }, params), auditContext);
    } catch (e) {
      log.error(e);
      return {};
    }
  }

  updateCommands(browser, customFn) {
    const commands = Object.getOwnPropertyNames(Object.getPrototypeOf(this)).filter(fnName => fnName !== 'constructor' && fnName !== 'updateCommands' && !fnName.startsWith('_'));
    commands.forEach(fnName => browser.addCommand(fnName, customFn || this[fnName].bind(this)));
  }

  async getMainThreadWorkBreakdown() {
    const result = await this._audit(_mainthreadWorkBreakdown.default);
    return result.details.items.map(({
      group,
      duration
    }) => ({
      group,
      duration
    }));
  }

  async getDiagnostics() {
    const result = await this._audit(_diagnostics.default);

    if (!Object.prototype.hasOwnProperty.call(result, 'details')) {
      return null;
    }

    return result.details.items[0];
  }

  async getMetrics() {
    const timeToFirstByte = await this._audit(_timeToFirstByte.default, {
      URL: this.url
    });
    const result = await this._audit(_metrics.default);
    const metrics = result.details.items[0] || {};
    return {
      estimatedInputLatency: metrics.estimatedInputLatency,
      timeToFirstByte: Math.round(timeToFirstByte.numericValue, 10),
      domContentLoaded: metrics.observedDomContentLoaded,
      firstVisualChange: metrics.observedFirstVisualChange,
      firstPaint: metrics.observedFirstPaint,
      firstContentfulPaint: metrics.firstContentfulPaint,
      firstMeaningfulPaint: metrics.firstMeaningfulPaint,
      lastVisualChange: metrics.observedLastVisualChange,
      firstCPUIdle: metrics.firstCPUIdle,
      firstInteractive: metrics.interactive,
      load: metrics.observedLoad,
      speedIndex: metrics.speedIndex
    };
  }

  async getPerformanceScore() {
    const {
      firstMeaningfulPaint,
      firstCPUIdle,
      firstInteractive,
      speedIndex,
      estimatedInputLatency
    } = await this.getMetrics();

    if (!firstMeaningfulPaint || !firstCPUIdle || !firstInteractive || !speedIndex || !estimatedInputLatency) {
      log.info('One or multiple required metrics couldn\'t be found, setting performance score to: null');
      return null;
    }

    const FMPScore = (0, _utils.quantileAtValue)(...METRIC_SCORING.FMP, firstMeaningfulPaint);
    const FIScore = (0, _utils.quantileAtValue)(...METRIC_SCORING.FI, firstCPUIdle);
    const CIScore = (0, _utils.quantileAtValue)(...METRIC_SCORING.CI, firstInteractive);
    const SIScore = (0, _utils.quantileAtValue)(...METRIC_SCORING.SI, speedIndex);
    const EILScore = (0, _utils.quantileAtValue)(...METRIC_SCORING.EIL, estimatedInputLatency);
    return (0, _weightedMean.default)([[FMPScore, WEIGHT_WITHIN_CATEGORY.FMP], [FIScore, WEIGHT_WITHIN_CATEGORY.FI], [CIScore, WEIGHT_WITHIN_CATEGORY.CI], [SIScore, WEIGHT_WITHIN_CATEGORY.SI], [EILScore, WEIGHT_WITHIN_CATEGORY.EIL]]);
  }

}

exports.default = Auditor;
